### ISSUE TITLE
Electricity Previous Accumulative Cost Override not updating

### ISSUE THREAD

**Issue by JRascagneres** [BottlecapDave/HomeAssistant-OctopusEnergy#671]
### Describe the bug

This issue seems a little intermittent but on some days it just doesn't update. As can be seen in the screenshot it didn't update but the consumption itself did. 

Cost override:
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/3264971/fe859e3a-ae7a-4ec2-813a-abd7f1b9e28c)

Consumption:
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/3264971/17bddb1e-e7b1-472b-a720-4da2776cb63f)

Here are logs around this:
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/3264971/fdc6fbd3-74ca-4614-9e09-1239a9c81251)

```
Update of sensor.octopus_energy_electricity_xxx_previous_accumulative_cost_override is taking over 10 seconds
Update of sensor.electricity_xxx_previous_accumulative_cost is taking over 10 seconds
Update of sensor.gas_xxx_previous_accumulative_consumption_kwh is taking over 10 seconds
Update of sensor.octopus_energy_electricity_xxx_export_previous_accumulative_cost is taking over 10 seconds
```

### Reproduction steps

Don't really do much... Just have override enabled and check the value.

### Expected behaviour

Update properly on a daily basis.

### Tariff Code

Normal: E-1R-AGILE-FLEX-22-11-25-D Override: E-1R-VAR-22-11-01-D

### Integration Version

9.2.1

### Home Assistant Version

2023.12.4

### Fresh Install?

Not specified

### Home Assistant Logs

Here are logs around this:
![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/3264971/fdc6fbd3-74ca-4614-9e09-1239a9c81251)

```
Update of sensor.octopus_energy_electricity_xxx_previous_accumulative_cost_override is taking over 10 seconds
Update of sensor.electricity_xxx_previous_accumulative_cost is taking over 10 seconds
Update of sensor.gas_xxx_previous_accumulative_consumption_kwh is taking over 10 seconds
Update of sensor.octopus_energy_electricity_xxx_export_previous_accumulative_cost is taking over 10 seconds
```

### Confirmation

- [X] I confirm that I cannot find my solution within the [documentation](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy)
- [X] I confirm that I cannot find my solution within the [FAQ](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/faq)

**Comment by BottlecapDave** [comment]
Hello. This could be related to https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/issues/670 if you have similar errors or it could be linked to the poor performance of the API. I have an override which seems to work. I'll see if anything jumps out at me while fixing the related issue, but theres not a lot to go on with this data unfortunately

**Comment by JRascagneres** [comment]
> Hello. This could be related to #670 if you have similar errors or it could be linked to the poor performance of the API. I have an override which seems to work. I'll see if anything jumps out at me while fixing the related issue, but theres not a lot to go on with this data unfortunately

I don't have any hard errors like that one

**Comment by JRascagneres** [comment]
@BottlecapDave Notably if I restart the integration it seems to pick it up. Wondering if something in the scheduling is broken in the new update code?

**Comment by BottlecapDave** [comment]
If the scheduling logic wasn't working I wouldn't expect those warnings, but there might be something not working if you're not getting any other related errors. I've added some additional logging which will be included in the next release

**Comment by JRascagneres** [comment]
I'm still getting this same issue. A restart of the integration fixes it every time and it immediately loads.

Previous accumulative consumption works fine.
Previous accumulative cost works fine.

It's just the previous accumulative cost override that doesn't seem to work.

**Comment by JRascagneres** [comment]
<img width="1016" alt="image" src="https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/3264971/057ab7bf-b966-44ec-9ad5-ca966f184c27">

Have this again and then the taking over 10 second stuff. I notice the octopus api failure there, did it fail to recover from that maybe? Unsure

**Comment by BottlecapDave** [comment]
Very odd. The raised error is not helping, but I'd have expected a bit more information :( I'll try and have a look when I get a chance.

**Comment by JRascagneres** [comment]
Thanks. This seems to happen on and off. Its been fine for the last couple days and then it was hit again today.

I realize this is an Octopus API error but this seems to be occurring which may be the cause of the issue? I can reach that endpoint though and, again, a restart of the integration almost immediately results in the data being pulled.

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/3264971/0d36150f-88a8-44bb-a890-d45f966cfc88)

Same pattern though, these two errors:

![image](https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/assets/3264971/bef523aa-adcb-4e85-935d-2f2a54fdf6b5)

**Comment by BottlecapDave** [comment]
@JRascagneres Sorry for the late response - been battling a winter illness. If this is till happening, the raw logs would be more useful as they show the timeline of things occuring. I think you're probably still being hit by slow API responses/timeout issues. I've increased the timeout slightly in the latest release, which may fix your issue.

**Comment by JRascagneres** [comment]
@BottlecapDave No problem! I also have had something. 
I've updated to the latest but it occurred last night, weirdly it still 'fixes' itself without fail when reloading the integration. 
For log wise what do you need? When you saw raw logs are these available somewhere?

**Comment by BottlecapDave** [comment]
The raw logs can be obtained by following the [faq](https://bottlecapdave.github.io/HomeAssistant-OctopusEnergy/faq/#how-do-i-increase-the-logs-for-the-integration) to turn on debugging. I probably need a couple of hours worth of logs.

**Comment by JRascagneres** [comment]
Thank you! I've enabled this now and I'll be back in touch when the issue is back. I assume it'll trigger tomorrow morning again. Is there a way I should send this? I think it'll have some potentially sensitive info in it.

**Comment by JRascagneres** [comment]
@BottlecapDave Hey, it happened again today. I've got the debug logs and I filtered it down to only show 'octopus' stuff. Can I send it to you some other way? Avoid uploading a bunch of serial numbers and things here?

**Comment by BottlecapDave** [comment]
Sorry to hear this. If you could send the logs to ha.octopus.energy@davidskendall.co.uk, then I'll take a look when I get a chance.

**Comment by BottlecapDave** [comment]
So I've had a look at the logs
- It looks like the entity is calling `Calculating previous electricity consumption cost override for 'xxx/yyy'...` every 30 seconds. The default for the entity should be every minute. Do you have an automation to try and update this entity every 30 seconds? Or do you somehow have multiple "override" entities registered in your instance? With this log being called at https://github.com/BottlecapDave/HomeAssistant-OctopusEnergy/blob/develop/custom_components/octopus_energy/electricity/previous_accumulative_cost_override.py#L132, it means the integration should be requesting the necessary data to perform the calculation.
- In the scenarios where the cache is being used, it looks like these are caused by timeout errors. I'm not sure if your HA instance is bottlenecked somewhere which is causing these issues. I'm currently targeting your specified tariff (with a different region) with no troubles.
- There is a bug in the "next refresh" logic where the next attempt time is in the past. This won't be helping with the backoffs, but not sure how it's got into this state. Perhaps the excessive calls mentioned in the first point is causing the update call to be cancelled early and therefore skipping on calculating the next attempt.  

I might need to add some more logging to try and work out where in the update it's failing, but at the moment nothing is jumping out at me.

**Comment by JRascagneres** [comment]
I don't have any automation around that. The only automation I have thats even related to the octopus integration is one that calculates the difference between the override and the 'normal' accumulative cost. I also don't have multiple override entities. I **DO** have an export meter though with an override? Would that cause it?

I don't think my HA instance is bottlenecked, I seem to have plenty of resources left and the network is fine. Would it be possible to have a user configurable timeout where I have it really large? Like a minute or something.

This bug seems interesting. Perhaps thats why it struggles and doesn't seem to recover unless I restart the integration. Restarting the integration gets me the update without fail.

**Comment by JRascagneres** [comment]
Its happening fairly consistently. It **always** works when restarting the integration. I think I might setup an automation to just reload the integration every day. Is there anything in your investigation that would explain why it always works on a restart. Is it something to do with the bug you mentioned?

**Comment by BottlecapDave** [comment]
Sorry for the delay in responding. I'll try and recreate your scenario locally with a template sensor to try and reproduce the results you're seeing (updates every 30 seconds instead of every 1 minute, staggered on failures).

Being able to configure the timeout will probably make things worse, as judging by the logs you provided, it looks like the updates are tripping up over themselves (HA doesn't seem to keep track if an update is in progress).

I'll probably look at moving the update logic to a background task so that it can't be influenced by the update entity call, which may be called as part of the template update.

**Comment by JRascagneres** [comment]
@BottlecapDave Thanks for your work on it, its really odd, like I say it always works on a restart so wondering what the difference is there.

**Comment by BottlecapDave** [comment]
It is odd. I think it's just not had a chance to trip over itself. The logs state it's retrieving the data, but not finishing the calculation for some reason.

**Comment by JRascagneres** [comment]
Interesting fair enough. Hopefully you're able to repro

**Comment by JRascagneres** [comment]
@BottlecapDave Have you been able to figure anything out, I'm still manually restarting once a day.

**Comment by BottlecapDave** [comment]
Sorry for the radio silence. I thought I fixed it, then I ended up breaking something else and with consumption data taking ages to populate it was taking a while to properly test.

Long story short, this should now be fixed in v10.1.0

**Comment by JRascagneres** [comment]
@BottlecapDave thank you! Not had this issue reoccur so seems fixed.
